Hygienic bug in macro (Ok function)

Similar to #16, but this time self_cell doesn't allow a custom `Ok` function:

```rs
use self_cell::self_cell;

// Remove this, and it compiles
pub fn Ok<T>(t: T) -> Result<T, ()> {
  Result::Ok(t)
}

#[derive(Debug, Eq, PartialEq)]
struct Ast<'a>(pub Vec<&'a str>);

self_cell!(
    struct AstCell {
        owner: String,

        #[covariant]
        dependent: Ast,
    }

    impl {Debug, Eq, PartialEq}
);

fn build_ast_cell(code: &str) -> AstCell {
  // Create owning String on stack.
  let pre_processed_code = code.trim().to_string();

  // Move String into AstCell, then build Ast inplace.
  AstCell::new(pre_processed_code, |code| {
    Ast(code.split(' ').filter(|word| word.len() > 1).collect())
  })
}

fn main() {
  let ast_cell = build_ast_cell("fox = cat + dog");

  println!("ast_cell -> {:?}", &ast_cell);
  println!("ast_cell.borrow_owner() -> {:?}", ast_cell.borrow_owner());
  println!(
    "ast_cell.borrow_dependent().0[1] -> {:?}",
    ast_cell.borrow_dependent().0[1]
  );
}
```

Compile error:

```
error[E0308]: mismatched types
  --> svc-gateway-host-run/src/e.rs:11:1
   |
11 | / self_cell!(
12 | |     struct AstCell {
13 | |         owner: String,
14 | |
...  |
19 | |     impl {Debug, Eq, PartialEq}
20 | | );
   | | ^
   | | |
   | | expected `Result<AstCell, Err>`, found `Result<AstCell, ()>`
   | |_expected this type parameter
   |   expected `Result<AstCell, Err>` because of return type
   |
   = note: expected enum `Result<_, Err>`
              found enum `Result<_, ()>`
   = note: this error originates in the macro `self_cell` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> svc-gateway-host-run/src/e.rs:11:1
   |
11 | / self_cell!(
12 | |     struct AstCell {
13 | |         owner: String,
14 | |
...  |
19 | |     impl {Debug, Eq, PartialEq}
20 | | );
   | | ^
   | | |
   | |_expected `Result<AstCell, (..., ...)>`, found `Result<AstCell, ()>`
   |   expected `Result<AstCell, (String, Err)>` because of return type
   |
   = note: expected enum `Result<_, (String, Err)>`
              found enum `Result<_, ()>`
   = note: this error originates in the macro `self_cell` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Context: was trying out [anyhow's `Ok` helper](https://docs.rs/anyhow/1.0.75/anyhow/fn.Ok.html)

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -29,6 +29,10 @@ jobs:
       run: |
         cd tests-extra
         cargo test --features=invalid_programs --verbose
+    - name: Run tests-extra no_std_lib
+      run: |
+        cd tests-extra/no_std_lib
+        cargo test --verbose
     - name: Run examples
       run: |
         cd examples
@@ -71,6 +75,9 @@ jobs:
         rustup component add miri
       shell: bash
 
+    - name: Run tests with nightly basic
+      run: |
+        cargo test --verbose
     - name: Run tests x86_64-unknown-linux-gnu
       run: |
         cargo miri test --verbose --target x86_64-unknown-linux-gnu

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -27,7 +27,7 @@
 //! impl NewStructName {
 //!     fn new(
 //!         owner: Owner,
-//!         dependent_builder: impl for<'a> std::ops::FnOnce(&'a Owner) -> Dependent<'a>
+//!         dependent_builder: impl for<'a> ::core::ops::FnOnce(&'a Owner) -> Dependent<'a>
 //!     ) -> NewStructName { ... }
 //!     fn borrow_owner<'a>(&'a self) -> &'a Owner { ... }
 //!     fn borrow_dependent<'a>(&'a self) -> &'a Dependent<'a> { ... }
@@ -191,21 +191,21 @@ pub mod unsafe_self_cell;
 /// ```ignore
 /// fn new(
 ///     owner: $Owner,
-///     dependent_builder: impl for<'a> std::ops::FnOnce(&'a $Owner) -> $Dependent<'a>
+///     dependent_builder: impl for<'a> ::core::ops::FnOnce(&'a $Owner) -> $Dependent<'a>
 /// ) -> Self
 /// ```
 ///
 /// ```ignore
 /// fn try_new<Err>(
 ///     owner: $Owner,
-///     dependent_builder: impl for<'a> std::ops::FnOnce(&'a $Owner) -> Result<$Dependent<'a>, Err>
+///     dependent_builder: impl for<'a> ::core::ops::FnOnce(&'a $Owner) -> Result<$Dependent<'a>, Err>
 /// ) -> Result<Self, Err>
 /// ```
 ///
 /// ```ignore
 /// fn try_new_or_recover<Err>(
 ///     owner: $Owner,
-///     dependent_builder: impl for<'a> std::ops::FnOnce(&'a $Owner) -> Result<$Dependent<'a>, Err>
+///     dependent_builder: impl for<'a> ::core::ops::FnOnce(&'a $Owner) -> Result<$Dependent<'a>, Err>
 /// ) -> Result<Self, ($Owner, Err)>
 /// ```
 ///
@@ -223,14 +223,14 @@ pub mod unsafe_self_cell;
 /// ```ignore
 /// fn with_dependent<'outer_fn, Ret>(
 ///     &'outer_fn self,
-///     func: impl for<'a> std::ops::FnOnce(&'a $Owner, &'outer_fn $Dependent<'a>
+///     func: impl for<'a> ::core::ops::FnOnce(&'a $Owner, &'outer_fn $Dependent<'a>
 /// ) -> Ret) -> Ret
 /// ```
 ///
 /// ```ignore
 /// fn with_dependent_mut<'outer_fn, Ret>(
 ///     &'outer_fn mut self,
-///     func: impl for<'a> std::ops::FnOnce(&'a $Owner, &'outer_fn mut $Dependent<'a>) -> Ret
+///     func: impl for<'a> ::core::ops::FnOnce(&'a $Owner, &'outer_fn mut $Dependent<'a>) -> Ret
 /// ) -> Ret
 /// ```
 ///
@@ -345,7 +345,7 @@ macro_rules! self_cell {
         /// remains valid for the lifetime of the constructed struct.
         $Vis fn new(
             owner: $Owner,
-            dependent_builder: impl for<'_q> std::ops::FnOnce(&'_q $Owner) -> $Dependent<'_q>
+            dependent_builder: impl for<'_q> ::core::ops::FnOnce(&'_q $Owner) -> $Dependent<'_q>
         ) -> Self {
             use ::core::ptr::NonNull;
 
@@ -359,7 +359,7 @@ macro_rules! self_cell {
                 // capture additional references in `dependent_builder` and then
                 // return them as part of Dependent. Eg. it should be impossible
                 // to express: '_q should outlive 'x here `fn
-                // bad<'_q>(outside_ref: &'_q String) -> impl for<'x> std::ops::FnOnce(&'x
+                // bad<'_q>(outside_ref: &'_q String) -> impl for<'x> ::core::ops::FnOnce(&'x
                 // Owner) -> Dependent<'x>`.
 
                 type JoinedCell<'_q $(, $OwnerLifetime)?> =
@@ -400,7 +400,7 @@ macro_rules! self_cell {
         $Vis fn try_new<Err>(
             owner: $Owner,
             dependent_builder:
-                impl for<'_q> std::ops::FnOnce(&'_q $Owner) -> ::core::result::Result<$Dependent<'_q>, Err>
+                impl for<'_q> ::core::ops::FnOnce(&'_q $Owner) -> ::core::result::Result<$Dependent<'_q>, Err>
         ) -> ::core::result::Result<Self, Err> {
             use ::core::ptr::NonNull;
 
@@ -449,7 +449,7 @@ macro_rules! self_cell {
         $Vis fn try_new_or_recover<Err>(
             owner: $Owner,
             dependent_builder:
-                impl for<'_q> std::ops::FnOnce(&'_q $Owner) -> ::core::result::Result<$Dependent<'_q>, Err>
+                impl for<'_q> ::core::ops::FnOnce(&'_q $Owner) -> ::core::result::Result<$Dependent<'_q>, Err>
         ) -> ::core::result::Result<Self, ($Owner, Err)> {
             use ::core::ptr::NonNull;
 
@@ -513,7 +513,7 @@ macro_rules! self_cell {
         /// Calls given closure `func` with a shared reference to dependent.
         $Vis fn with_dependent<'outer_fn, Ret>(
             &'outer_fn self,
-            func: impl for<'_q> std::ops::FnOnce(&'_q $Owner, &'outer_fn $Dependent<'_q>
+            func: impl for<'_q> ::core::ops::FnOnce(&'_q $Owner, &'outer_fn $Dependent<'_q>
         ) -> Ret) -> Ret {
             unsafe {
                 func(
@@ -526,7 +526,7 @@ macro_rules! self_cell {
         /// Calls given closure `func` with an unique reference to dependent.
         $Vis fn with_dependent_mut<'outer_fn, Ret>(
             &'outer_fn mut self,
-            func: impl for<'_q> std::ops::FnOnce(&'_q $Owner, &'outer_fn mut $Dependent<'_q>) -> Ret
+            func: impl for<'_q> ::core::ops::FnOnce(&'_q $Owner, &'outer_fn mut $Dependent<'_q>) -> Ret
         ) -> Ret {
             let (owner, dependent) = unsafe {
                     self.unsafe_self_cell.borrow_mut()

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -27,7 +27,7 @@
 //! impl NewStructName {
 //!     fn new(
 //!         owner: Owner,
-//!         dependent_builder: impl for<'a> FnOnce(&'a Owner) -> Dependent<'a>
+//!         dependent_builder: impl for<'a> std::ops::FnOnce(&'a Owner) -> Dependent<'a>
 //!     ) -> NewStructName { ... }
 //!     fn borrow_owner<'a>(&'a self) -> &'a Owner { ... }
 //!     fn borrow_dependent<'a>(&'a self) -> &'a Dependent<'a> { ... }
@@ -191,21 +191,21 @@ pub mod unsafe_self_cell;
 /// ```ignore
 /// fn new(
 ///     owner: $Owner,
-///     dependent_builder: impl for<'a> FnOnce(&'a $Owner) -> $Dependent<'a>
+///     dependent_builder: impl for<'a> std::ops::FnOnce(&'a $Owner) -> $Dependent<'a>
 /// ) -> Self
 /// ```
 ///
 /// ```ignore
 /// fn try_new<Err>(
 ///     owner: $Owner,
-///     dependent_builder: impl for<'a> FnOnce(&'a $Owner) -> Result<$Dependent<'a>, Err>
+///     dependent_builder: impl for<'a> std::ops::FnOnce(&'a $Owner) -> Result<$Dependent<'a>, Err>
 /// ) -> Result<Self, Err>
 /// ```
 ///
 /// ```ignore
 /// fn try_new_or_recover<Err>(
 ///     owner: $Owner,
-///     dependent_builder: impl for<'a> FnOnce(&'a $Owner) -> Result<$Dependent<'a>, Err>
+///     dependent_builder: impl for<'a> std::ops::FnOnce(&'a $Owner) -> Result<$Dependent<'a>, Err>
 /// ) -> Result<Self, ($Owner, Err)>
 /// ```
 ///
@@ -223,14 +223,14 @@ pub mod unsafe_self_cell;
 /// ```ignore
 /// fn with_dependent<'outer_fn, Ret>(
 ///     &'outer_fn self,
-///     func: impl for<'a> FnOnce(&'a $Owner, &'outer_fn $Dependent<'a>
+///     func: impl for<'a> std::ops::FnOnce(&'a $Owner, &'outer_fn $Dependent<'a>
 /// ) -> Ret) -> Ret
 /// ```
 ///
 /// ```ignore
 /// fn with_dependent_mut<'outer_fn, Ret>(
 ///     &'outer_fn mut self,
-///     func: impl for<'a> FnOnce(&'a $Owner, &'outer_fn mut $Dependent<'a>) -> Ret
+///     func: impl for<'a> std::ops::FnOnce(&'a $Owner, &'outer_fn mut $Dependent<'a>) -> Ret
 /// ) -> Ret
 /// ```
 ///
@@ -345,9 +345,9 @@ macro_rules! self_cell {
         /// remains valid for the lifetime of the constructed struct.
         $Vis fn new(
             owner: $Owner,
-            dependent_builder: impl for<'_q> FnOnce(&'_q $Owner) -> $Dependent<'_q>
+            dependent_builder: impl for<'_q> std::ops::FnOnce(&'_q $Owner) -> $Dependent<'_q>
         ) -> Self {
-            use core::ptr::NonNull;
+            use ::core::ptr::NonNull;
 
             unsafe {
                 // All this has to happen here, because there is not good way
@@ -359,7 +359,7 @@ macro_rules! self_cell {
                 // capture additional references in `dependent_builder` and then
                 // return them as part of Dependent. Eg. it should be impossible
                 // to express: '_q should outlive 'x here `fn
-                // bad<'_q>(outside_ref: &'_q String) -> impl for<'x> FnOnce(&'x
+                // bad<'_q>(outside_ref: &'_q String) -> impl for<'x> std::ops::FnOnce(&'x
                 // Owner) -> Dependent<'x>`.
 
                 type JoinedCell<'_q $(, $OwnerLifetime)?> =
@@ -383,7 +383,7 @@ macro_rules! self_cell {
 
                 // Initialize dependent with owner reference in final place.
                 dependent_ptr.write(dependent_builder(&*owner_ptr));
-                core::mem::forget(drop_guard);
+                ::core::mem::forget(drop_guard);
 
                 Self {
                     unsafe_self_cell: $crate::unsafe_self_cell::UnsafeSelfCell::new(
@@ -400,9 +400,9 @@ macro_rules! self_cell {
         $Vis fn try_new<Err>(
             owner: $Owner,
             dependent_builder:
-                impl for<'_q> FnOnce(&'_q $Owner) -> core::result::Result<$Dependent<'_q>, Err>
-        ) -> core::result::Result<Self, Err> {
-            use core::ptr::NonNull;
+                impl for<'_q> std::ops::FnOnce(&'_q $Owner) -> ::core::result::Result<$Dependent<'_q>, Err>
+        ) -> ::core::result::Result<Self, Err> {
+            use ::core::ptr::NonNull;
 
             unsafe {
                 // See fn new for more explanation.
@@ -427,18 +427,18 @@ macro_rules! self_cell {
                     $crate::unsafe_self_cell::OwnerAndCellDropGuard::new(joined_ptr);
 
                 match dependent_builder(&*owner_ptr) {
-                    Ok(dependent) => {
+                    ::core::result::Result::Ok(dependent) => {
                         dependent_ptr.write(dependent);
-                        core::mem::forget(drop_guard);
+                        ::core::mem::forget(drop_guard);
 
-                        Ok(Self {
+                        ::core::result::Result::Ok(Self {
                             unsafe_self_cell: $crate::unsafe_self_cell::UnsafeSelfCell::new(
                                 joined_void_ptr,
                             ),
                             $(owner_marker: $crate::_covariant_owner_marker_ctor!($OwnerLifetime) ,)?
                         })
                     }
-                    Err(err) => Err(err)
+                    ::core::result::Result::Err(err) => ::core::result::Result::Err(err)
                 }
             }
         }
@@ -449,9 +449,9 @@ macro_rules! self_cell {
         $Vis fn try_new_or_recover<Err>(
             owner: $Owner,
             dependent_builder:
-                impl for<'_q> FnOnce(&'_q $Owner) -> core::result::Result<$Dependent<'_q>, Err>
-        ) -> core::result::Result<Self, ($Owner, Err)> {
-            use core::ptr::NonNull;
+                impl for<'_q> std::ops::FnOnce(&'_q $Owner) -> ::core::result::Result<$Dependent<'_q>, Err>
+        ) -> ::core::result::Result<Self, ($Owner, Err)> {
+            use ::core::ptr::NonNull;
 
             unsafe {
                 // See fn new for more explanation.
@@ -476,30 +476,30 @@ macro_rules! self_cell {
                     $crate::unsafe_self_cell::OwnerAndCellDropGuard::new(joined_ptr);
 
                 match dependent_builder(&*owner_ptr) {
-                    Ok(dependent) => {
+                    ::core::result::Result::Ok(dependent) => {
                         dependent_ptr.write(dependent);
-                        core::mem::forget(drop_guard);
+                        ::core::mem::forget(drop_guard);
 
-                        Ok(Self {
+                        ::core::result::Result::Ok(Self {
                             unsafe_self_cell: $crate::unsafe_self_cell::UnsafeSelfCell::new(
                                 joined_void_ptr,
                             ),
                             $(owner_marker: $crate::_covariant_owner_marker_ctor!($OwnerLifetime) ,)?
                         })
                     }
-                    Err(err) => {
+                    ::core::result::Result::Err(err) => {
                         // In contrast to into_owner ptr::read, here no dependent
                         // ever existed in this function and so we are sure its
                         // drop impl can't access owner after the read.
                         // And err can't return a reference to owner.
-                        let owner_on_err = core::ptr::read(owner_ptr);
+                        let owner_on_err = ::core::ptr::read(owner_ptr);
 
                         // Allowing drop_guard to finish would let it double free owner.
                         // So we dealloc the JoinedCell here manually.
-                        core::mem::forget(drop_guard);
+                        ::core::mem::forget(drop_guard);
                         $crate::alloc::alloc::dealloc(joined_void_ptr.as_ptr(), layout);
 
-                        Err((owner_on_err, err))
+                        ::core::result::Result::Err((owner_on_err, err))
                     }
                 }
             }
@@ -513,7 +513,7 @@ macro_rules! self_cell {
         /// Calls given closure `func` with a shared reference to dependent.
         $Vis fn with_dependent<'outer_fn, Ret>(
             &'outer_fn self,
-            func: impl for<'_q> FnOnce(&'_q $Owner, &'outer_fn $Dependent<'_q>
+            func: impl for<'_q> std::ops::FnOnce(&'_q $Owner, &'outer_fn $Dependent<'_q>
         ) -> Ret) -> Ret {
             unsafe {
                 func(
@@ -526,7 +526,7 @@ macro_rules! self_cell {
         /// Calls given closure `func` with an unique reference to dependent.
         $Vis fn with_dependent_mut<'outer_fn, Ret>(
             &'outer_fn mut self,
-            func: impl for<'_q> FnOnce(&'_q $Owner, &'outer_fn mut $Dependent<'_q>) -> Ret
+            func: impl for<'_q> std::ops::FnOnce(&'_q $Owner, &'outer_fn mut $Dependent<'_q>) -> Ret
         ) -> Ret {
             let (owner, dependent) = unsafe {
                     self.unsafe_self_cell.borrow_mut()
@@ -540,7 +540,7 @@ macro_rules! self_cell {
         /// Consumes `self` and returns the the owner.
         $Vis fn into_owner(self) -> $Owner {
             // This is only safe to do with repr(transparent).
-            let unsafe_self_cell = unsafe { core::mem::transmute::<
+            let unsafe_self_cell = unsafe { ::core::mem::transmute::<
                 Self,
                 $crate::unsafe_self_cell::UnsafeSelfCell<
                     $StructName$(<$OwnerLifetime>)?,
@@ -602,14 +602,14 @@ macro_rules! _covariant_owner_marker {
     (covariant, $OwnerLifetime:lifetime) => {
         // Ensure that contravariant owners don't imply covariance
         // over the dependent. See issue https://github.com/Voultapher/self_cell/issues/18
-        core::marker::PhantomData<&$OwnerLifetime ()>
+        ::core::marker::PhantomData<&$OwnerLifetime ()>
     };
     (not_covariant, $OwnerLifetime:lifetime) => {
         // See the discussion in https://github.com/Voultapher/self_cell/pull/29
         //
         // If the dependent is non_covariant, mark the owner as invariant over its
         // lifetime. Otherwise unsound use is possible.
-        core::marker::PhantomData<fn(&$OwnerLifetime ()) -> &$OwnerLifetime ()>
+        ::core::marker::PhantomData<fn(&$OwnerLifetime ()) -> &$OwnerLifetime ()>
     };
     ($x:ident, $OwnerLifetime:lifetime) => {
         compile_error!("This macro only accepts `covariant` or `not_covariant`");
@@ -621,7 +621,7 @@ macro_rules! _covariant_owner_marker {
 macro_rules! _covariant_owner_marker_ctor {
     ($OwnerLifetime:lifetime) => {
         // Helper to optionally expand into PhantomData for construction.
-        core::marker::PhantomData
+        ::core::marker::PhantomData
     };
 }
 
@@ -629,11 +629,11 @@ macro_rules! _covariant_owner_marker_ctor {
 #[macro_export]
 macro_rules! _impl_automatic_derive {
     (Debug, $StructName:ident) => {
-        impl core::fmt::Debug for $StructName {
+        impl ::core::fmt::Debug for $StructName {
             fn fmt(
                 &self,
-                fmt: &mut core::fmt::Formatter,
-            ) -> core::result::Result<(), core::fmt::Error> {
+                fmt: &mut ::core::fmt::Formatter,
+            ) -> ::core::result::Result<(), ::core::fmt::Error> {
                 self.with_dependent(|owner, dependent| {
                     fmt.debug_struct(stringify!($StructName))
                         .field("owner", owner)
@@ -644,7 +644,7 @@ macro_rules! _impl_automatic_derive {
         }
     };
     (PartialEq, $StructName:ident) => {
-        impl core::cmp::PartialEq for $StructName {
+        impl ::core::cmp::PartialEq for $StructName {
             fn eq(&self, other: &Self) -> bool {
                 *self.borrow_owner() == *other.borrow_owner()
             }
@@ -652,11 +652,11 @@ macro_rules! _impl_automatic_derive {
     };
     (Eq, $StructName:ident) => {
         // TODO this should only be allowed if owner is Eq.
-        impl core::cmp::Eq for $StructName {}
+        impl ::core::cmp::Eq for $StructName {}
     };
     (Hash, $StructName:ident) => {
-        impl core::hash::Hash for $StructName {
-            fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        impl ::core::hash::Hash for $StructName {
+            fn hash<H: ::core::hash::Hasher>(&self, state: &mut H) {
                 self.borrow_owner().hash(state);
             }
         }

--- a/src/unsafe_self_cell.rs
+++ b/src/unsafe_self_cell.rs
@@ -1,3 +1,5 @@
+#![allow(clippy::needless_lifetimes)]
+
 use core::marker::PhantomData;
 use core::mem;
 use core::ptr::{drop_in_place, read, NonNull};

--- a/tests-extra/no_std_lib/Cargo.toml
+++ b/tests-extra/no_std_lib/Cargo.toml
@@ -1,0 +1,10 @@
+[package]
+name = "no_std_lib"
+version = "0.1.0"
+authors = ["Lukas Bergdoll <lukas.bergdoll@gmail.com>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+self_cell = { path = "../.." }

--- a/tests-extra/no_std_lib/src/lib.rs
+++ b/tests-extra/no_std_lib/src/lib.rs
@@ -1,0 +1,59 @@
+#![no_std]
+
+use self_cell::self_cell;
+
+// Not using alloc is on purpose, self_cell should also work in such scenarios.
+
+const SCRATCH_REGION: [u8; 4096] = [0u8; 4096];
+
+#[derive(Eq, PartialEq)]
+struct StaticString {
+    region: &'static [u8],
+}
+
+const MAX_NODES: usize = 8;
+
+#[derive(Eq, PartialEq)]
+struct Ast<'a>([Option<&'a [u8]>; MAX_NODES]);
+
+self_cell!(
+    struct AstCell {
+        owner: StaticString,
+
+        #[covariant]
+        dependent: Ast,
+    }
+
+    impl {Eq, PartialEq}
+);
+
+fn build_ast_cell() -> AstCell {
+    // Yes in a static setting self_cell is not terribly useful, this could be solved differently.
+    // This is only a test.
+    let pre_processed_code = StaticString {
+        region: &SCRATCH_REGION[4000..4015],
+    };
+
+    AstCell::new(pre_processed_code, |code| {
+        let mut ast_nodes = [None; MAX_NODES];
+        ast_nodes[0] = Some(&code.region[3..7]);
+        ast_nodes[1] = Some(&code.region[10..12]);
+
+        Ast(ast_nodes)
+    })
+}
+
+#[test]
+fn self_cell_works_in_no_std_env() {
+    let ast_cell = build_ast_cell();
+    assert_eq!(ast_cell.borrow_owner().region.len(), 15);
+    assert_eq!(
+        ast_cell
+            .borrow_dependent()
+            .0
+            .iter()
+            .filter(|val| val.is_some())
+            .count(),
+        2
+    );
+}

--- a/tests/self_cell.rs
+++ b/tests/self_cell.rs
@@ -94,7 +94,7 @@ struct FnOnce<T> {
 mod core {
     pub mod ptr {
         #[allow(dead_code)]
-        pub const unsafe fn read<T>(_src: *const T) -> T {
+        pub unsafe fn read<T>(_src: *const T) -> T {
             panic!("Wrong ptr::read function called")
         }
     }

--- a/tests/self_cell.rs
+++ b/tests/self_cell.rs
@@ -55,7 +55,7 @@ impl PackedAst {
         self.ast_cell.borrow_owner()
     }
 
-    fn with_ast<'o>(&'o self, func: impl for<'a> FnOnce(&'a String, &'o Ast<'a>)) {
+    fn with_ast<'o>(&'o self, func: impl for<'a> std::ops::FnOnce(&'a String, &'o Ast<'a>)) {
         self.ast_cell.with_dependent(func)
     }
 
@@ -71,6 +71,32 @@ fn assert_with_ast(packed_ast: &PackedAst, expected_ast: &Ast) {
         visited = true;
     });
     assert!(visited);
+}
+
+// These are defined visible for the whole module. Because incorrect usage could affect many
+// components it is not isolated to a specific test.
+#[allow(non_snake_case, dead_code)]
+fn Ok<T>(_: T) -> ! {
+    panic!("Wrong Ok function called")
+}
+
+#[allow(non_snake_case, dead_code)]
+fn Err<T>(_: T) -> ! {
+    panic!("Wrong Err function called")
+}
+
+#[allow(dead_code)]
+struct FnOnce<T> {
+    marker: T,
+}
+
+mod core {
+    pub mod ptr {
+        #[allow(dead_code)]
+        pub const unsafe fn read<T>(_src: *const T) -> T {
+            panic!("Wrong ptr::read function called")
+        }
+    }
 }
 
 #[test]
@@ -139,7 +165,9 @@ fn failable_constructor_success() {
     let expected_ast = Ast::from(&owner);
 
     let ast_cell_result: Result<PackedAstCell, i32> =
-        PackedAstCell::try_new(owner.clone(), |owner| Ok(Ast::from(owner)));
+        PackedAstCell::try_new(owner.clone(), |owner| {
+            std::result::Result::Ok(Ast::from(owner))
+        });
 
     assert!(ast_cell_result.is_ok());
 
@@ -153,7 +181,7 @@ fn failable_constructor_fail() {
     let owner = String::from("This string is no trout");
 
     let ast_cell_result: Result<PackedAstCell, i32> =
-        PackedAstCell::try_new(owner.clone(), |_owner| Err(22));
+        PackedAstCell::try_new(owner.clone(), |_owner| std::result::Result::Err(22));
 
     assert!(ast_cell_result.is_err());
 
@@ -504,14 +532,15 @@ fn try_new_or_recover() {
 
     // bad path
     let (input, err) =
-        PackedAstCell::try_new_or_recover(original_input.clone(), |_| Err(-1)).unwrap_err();
+        PackedAstCell::try_new_or_recover(original_input.clone(), |_| std::result::Result::Err(-1))
+            .unwrap_err();
 
     assert_eq!(original_input, input);
     assert_eq!(err, -1);
 
     // happy path
     let cell = PackedAstCell::try_new_or_recover(original_input.clone(), |o| -> Result<_, ()> {
-        Ok(o.into())
+        std::result::Result::Ok(o.into())
     })
     .unwrap();
     assert_eq!(cell.borrow_owner(), &original_input);
@@ -567,14 +596,14 @@ fn zero_size_cell() {
 
     assert!(
         catch_unwind(|| ZeroSizeCell::try_new((), |_| -> Result<_, i32> {
-            Ok(ZeroSizeRef(PhantomData))
+            std::result::Result::Ok(ZeroSizeRef(PhantomData))
         }))
         .is_err()
     );
 
     assert!(catch_unwind(
         || ZeroSizeCell::try_new_or_recover((), |_| -> Result<_, i32> {
-            Ok(ZeroSizeRef(PhantomData))
+            std::result::Result::Ok(ZeroSizeRef(PhantomData))
         })
     )
     .is_err());

--- a/tests/self_cell.rs
+++ b/tests/self_cell.rs
@@ -16,7 +16,7 @@ use self_cell::self_cell;
 pub struct Ast<'input>(pub Vec<&'input str>);
 
 impl<'x> From<&'x String> for Ast<'x> {
-    fn from<'a>(body: &'a String) -> Ast<'a> {
+    fn from(body: &String) -> Ast<'_> {
         Ast(vec![&body[2..5], &body[1..3]])
     }
 }
@@ -51,7 +51,7 @@ impl PackedAst {
         }
     }
 
-    fn get_body<'a>(&'a self) -> &'a String {
+    fn get_body(&self) -> &String {
         self.ast_cell.borrow_owner()
     }
 
@@ -59,6 +59,7 @@ impl PackedAst {
         self.ast_cell.with_dependent(func)
     }
 
+    #[allow(clippy::needless_lifetimes)]
     fn get_ast<'a>(&'a self) -> &'a Ast<'a> {
         self.ast_cell.borrow_dependent()
     }
@@ -141,7 +142,7 @@ fn parse_ast() {
 
 fn make_ast_with_stripped_body(body: &str) -> PackedAst {
     // This is created on the stack.
-    let stripped_body = body.replace("\n", "");
+    let stripped_body = body.replace('\n', "");
     // Return Ast built from moved body, no lifetime hassle.
     PackedAst::new(stripped_body)
 }
@@ -149,7 +150,7 @@ fn make_ast_with_stripped_body(body: &str) -> PackedAst {
 #[test]
 fn return_self_ref_struct() {
     let body = String::from("a\nb\nc\ndef");
-    let expected_body = body.replace("\n", "");
+    let expected_body = body.replace('\n', "");
 
     // expected_ast is on the stack and lifetime dependent on body.
     let expected_ast = Ast::from(&expected_body);
@@ -241,7 +242,8 @@ fn catch_panic_in_from() {
 
     impl<'a> PanicCtor<'a> {
         fn new(_: &'a Owner) -> Self {
-            let _stack_vec = vec![23, 44, 5];
+            let stack_vec = vec![23, 44, 5];
+            println!("{stack_vec:?}");
             panic!()
         }
     }
@@ -261,7 +263,7 @@ fn catch_panic_in_from() {
     let owner = Owner("This string is no trout".into());
 
     let ast_cell_result = NoLeakCell::try_new(owner.clone(), |owner| {
-        catch_unwind(|| PanicCtor::new(&owner))
+        catch_unwind(|| PanicCtor::new(owner))
     });
     assert!(ast_cell_result.is_err());
 }

--- a/tests/self_cell.rs
+++ b/tests/self_cell.rs
@@ -243,7 +243,7 @@ fn catch_panic_in_from() {
     impl<'a> PanicCtor<'a> {
         fn new(_: &'a Owner) -> Self {
             let stack_vec = vec![23, 44, 5];
-            println!("{stack_vec:?}");
+            println!("{:?}", stack_vec);
             panic!()
         }
     }

